README.md
# NAME

YCGL::Lite - Yokoda Common General Library lite.

# SYNOPSIS

    use strict;
    use warnings;
    use feature qw(say);
    use YCGL::Lite;

    my $ycgl = YCGL::Lite->new();

    ## Current timestamp
    my $current_timestamp = $ycgl->times->current_timestamp();
    say $current_timestamp; # 2016-01-01 00:00:00
    my $today = $ycgl->times->today();
    say $today; # 2016-01-01

    my $date_one_month_ago = $ycgl->times->date_add($today,0,-1,0);
    say $date_one_month_ago; # 2015-12-01
    my $date_diff = $ycgl->times->date_diff($today,'2016-01-05');
    say $date_diff; # 4

    ## Mail
    my $mail_data = {
       to => [{name => 'to1',addr =>'to1@example.com'}],
       cc => [{name => 'cc1', addr => 'cc1@example.com'}],
       bcc => [{name => 'bcc', addr => 'bcc1@example.com'}],
       from => {name => 'myname', addr => 'myname@example.com'},
       mta_addr => '127.0.0.1',
       subject => 'mail example.',
       content => 'This is example.',
       content_type => 'text/plain'
    };

    $ycgl->mail->send_mail($mail_data);

    ## HTTP, HTTPS client
    my $result_http = $ycgl->http_client->get('http://example.com');
    my $result_https = $ycgl->http_client->post('https://example.com',{username => 'myname', password => 'password'});

    ## Data interchange
    # JSON
    my $json_text = $ycgl->data_conv->perl_to_json({name => 'myname', age => 20});
    my $perl_from_json = $ycgl->data_conv->json_to_perl($json_text);

    # CSV
    my $csv_text = $ycgl->data_conv->perl_to_csv([['name','age'],['myname',20],['yourname',21]]);
    my $perl_from_csv = $ycgl->data_conv->csv_to_perl($csv_text);
    my $perl_from_csvfile = $ycgl->data_conv->csvfile_to_perl('./mycsvfile.csv');
    $ycgl->data_conv->perl_to_csvfile($perl_from_csvfile,'./mycsvoutput.csv');

    # XML
    my $xml_records = {
      meta_data => [{source_name => ['MYSOURCE']}],
      records =>
      [
        {
          data_seq => 1,
          pname => 'myname',
          age => 20
        },
        {
          data_seq => 2,
          pname => 'yourname',
          age => 21
        }
      ]
    };
    my $xml_text = $ycgl->data_conv->perl_to_xml(
      $xml_records,
      {ForceArray => 1,RootName => 'ROOTNAME', XMLDecl => 1} # option
    );
    my $perl_from_xml = $ycgl->data_conv->xml_to_perl($xml_text);

    # MessagePack
    my $msgpack_text = $ycgl->data_conv->perl_to_msgpack({name => 'myname', age => 20});
    my $perl_from_msgpack = $ycgl->data_conv->msgpack_to_perl($msgpack_text);

    ## Parallel Processing
    my $data_parallel_1 = [{val_1 => 1, val_2 => 2},{val_1 => 10, val_2 => 20}];
    my $my_sub_1 = sub {my $inputs = shift; return($inputs{val_1} + $inputs{val_2});};
    my @results = $ycgl->parallel->do_with_result($data_parallel_1, $my_sub_1, 10);

    my $data_parallel_2 = ['http://example.com/1','http://example.com/2'];
    my $my_sub_2 = sub {my $url = shift; $ycgl->http_client->get($url);};
    $ycgl->parallel->do_without_result($data_parallel_2, $my_sub_2, 20);

    ## MapReduce
    # generating data
    my $data_map_reduce;
    for(0 .. 20){
        my $tmp_data;
        for(0 .. 10000){
            push(@$tmp_data,rand(10000));
        }
        push(@$data_map_reduce,$tmp_data);
    }

    # mapper code
    my $mapper = sub {
        my $input = shift;
        my $sum = 0;
        my $num = $#$input + 1;
        for(0 .. $#$input){
            $sum += $input->[$_];
        }
        my $avg = $sum / $num;
        return({avg => $avg, sum => $sum});
    };

    # reducer code
    my $reducer = sub {
        my $input = shift;
        my $sum = 0;
        my $avg = 0;
        my $num = $#$input + 1;
        for(0 .. $#$input){
            $sum += $input->[$_]->{sum};
            $avg += $input->[$_]->{avg};
        }
        $avg = $avg / $num;
        return({avg => $avg, sum => $sum});
    };

    # do MapReduce
    my $result = $ycgl->parallel->map_reduce(
        $data,
        $mapper,
        $reducer,
        10
       );
    print "SUM: $result->{sum}\nAVG: $result->{avg}\n";

    ## Plack Server
    # static file server
    my $public_dir = './public/';
    $ycgl->plack->plackup_static($public_dir);

    # eval server (dangerous)
    my $eval_path = '/mysecret_eval_path';
    $ycgl->plack->plackup_eval($eval_path);

    my $data_map_reduce_real;
    for(0 .. 2){
        my $tmp_data;
        for(0 .. 10000){
            push(@$tmp_data,rand(10000));
        }
        push(@$data_map_reduce_real,[$tmp_data,'http://localhost:5000/mysecret_eval_path']);
    }

    # MapReduce by using remote eval servers
    my $real_mapr_result = $ycgl->parallel->map_reduce(
        $data_map_reduce_real,
        $mapper,
        $reducer,
        3,
        {remote => 1}
        );

# DESCRIPTION

YCGL::Lite has commonly and generally used functions.

# LICENSE

Copyright (C) Toshiaki Yokoda.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Toshiaki Yokoda <adokoy001@gmail.com>


Changes
Revision history for Perl extension YCGL-Lite

{{$NEXT}}

    - original version



lib/YCGL/Lite/Plack.pm
package YCGL::Lite::Plack;
use strict;
use warnings;
use Safe;
use Mouse;
use Plack::Runner;
use Plack::App::File;
use Plack::Request;

sub plackup_static{
    my $self = shift;
    my $public_dir = shift;
    my $app = Plack::App::File->new(root => $public_dir)->to_app;
    my $runner = Plack::Runner->new;
    $runner->run($app);
}

sub plackup_eval {
    my $self = shift;
    my $path = shift;
    my $app = sub {
	my $env = shift;
	my $req = Plack::Request->new($env);
	my $response = {
	    $path => sub {
		my $json_req = $req->content //
		    return [400,['Content-Type' => 'text/plain'],['Content body required.']];
		my $perl_req = JSON::decode_json($json_req) //
		    return [400,['Content-Type' => 'text/plain'],['Valid JSON required']];
		my $data = $perl_req->{data};
		my $code_text = $perl_req->{code};
		my $code_ref;
		eval('$code_ref = sub '.$code_text.';');
		my $result = $code_ref->($data);
		return [200,['Content-Type' => 'application/json'],[JSON::encode_json({result => $result})]];
	    }
	   };
	if(defined($response->{$env->{PATH_INFO}})){
	    return $response->{$env->{PATH_INFO}}->();
	}else{
	    return [404,['Content-Type' => 'text/plain'],['Not Found']];
	}
    };
    my $runner = Plack::Runner->new;
    $runner->run($app);
}

__PACKAGE__->meta->make_immutable();

1;


lib/YCGL/Lite/Times.pm
package YCGL::Lite::Times;
use strict;
use warnings;
use Time::HiRes;
use Date::Calc;
use Mouse;

has 'internal_timer' => ( is => 'rw', isa => 'Num', default => 0);

sub timer_start {
    my $self = shift;
    my $current_time = Time::HiRes::time;
    $self->internal_timer($current_time);
};

sub timer_elapsed {
    my $self = shift;
    my $diff = Time::HiRes::time - $self->internal_timer;
    return($diff);
};

sub timer_reset {
    my $self = shift;
    $self->internal_timer(0);
}

# get today date style
sub today{
    my $self = shift;
    my ($date_year,$date_month,$date_day) = Date::Calc::Today();
    my $date_result = sprintf(
	"%04d-%02d-%02d",
	$date_year,
	$date_month,
	$date_day
       );
    return($date_result);
}

# get current timestamp
sub current_timestamp{
    my $self = shift;
    my ($date_year,$date_month,$date_day,$date_hour,$date_min,$date_sec) = Date::Calc::Today_and_Now();
    my $date_result = sprintf(
	"%04d-%02d-%02d %02d:%02d:%02d",
	$date_year,
	$date_month,
	$date_day,
	$date_hour,
	$date_min,
	$date_sec
       );
    return($date_result);
}

sub date_add{
    my $self = shift;
    my $date = shift;
    my $diff_year = shift;
    my $diff_month = shift;
    my $diff_day = shift;
    my ($date_year,$date_month,$date_day) = _split_date($date);
    my @date_result_array = Date::Calc::Add_Delta_YMD(
	$date_year, $date_month, $date_day,
	$diff_year, $diff_month, $diff_day);
    my $date_result = sprintf(
	"%04d-%02d-%02d",
	$date_result_array[0],
	$date_result_array[1],
	$date_result_array[2]
       );
    return($date_result);
}

sub date_diff{
    my $self = shift;
    my $date_1 = shift;
    my $date_2 = shift;
    my ($date_year_1,$date_month_1,$date_day_1) = _split_date($date_1);
    my ($date_year_2,$date_month_2,$date_day_2) = _split_date($date_2);
    my $diff = Date::Calc::Delta_Days(
	$date_year_1,$date_month_1,$date_day_1,
	$date_year_2,$date_month_2,$date_day_2
       );
    return($diff);
}

sub split_date {
    my $self = shift;
    my $date = shift;
    my ($date_year,$date_month,$date_day);
    if($date =~ /([0-9]{4})[\-\/]{0,1}([0-9]{2})[\-\/]{0,1}([0-9]{2})/){
	$date_year = $1;
	$date_month = $2;
	$date_day = $3;
    }
    return($date_year,$date_month,$date_day);
}

sub split_timestamp {
    my $self = shift;
    my $timestamp = shift;
    my ($date_year,$date_month,$date_day,$hour,$min,$sec);
    if($timestamp =~ /([0-9]{4})[\-\/]{0,1}([0-9]{2})[\-\/]{0,1}([0-9]{2}) ([0-9]{2}):([0-9]{2}):([0-9]{2})/){
	$date_year = $1;
	$date_month = $2;
	$date_day = $3;
	$hour = $4;
	$min = $5;
	$sec = $6;
    }
    return($date_year,$date_month,$date_day,$hour,$min,$sec);
}


# internal function
sub _split_date {
    my $date = shift;
    my ($date_year,$date_month,$date_day);
    if($date =~ /([0-9]{4})[\-\/]{0,1}([0-9]{2})[\-\/]{0,1}([0-9]{2})/){
	$date_year = $1;
	$date_month = $2;
	$date_day = $3;
    }
    return($date_year,$date_month,$date_day);
}

__PACKAGE__->meta->make_immutable();

1;


lib/YCGL/Lite/Mailer.pm
package YCGL::Lite::Mailer;
use strict;
use warnings;
use utf8;
use Mouse;
use Encode qw(from_to encode);
use Net::SMTP;


sub send_mail{
    my $self = shift;
    my $mail_config = shift;
    my $mail_list_ref = $mail_config->{to};
    my $from_addr = $mail_config->{from};
    my $subject = $mail_config->{subject};
    my $content = $mail_config->{content};
    my $mta_addr = $mail_config->{mta_addr};
    my $content_type = $mail_config->{content_type};
    my $mail_list_ref_cc = $mail_config->{cc};
    my $mail_list_ref_bcc = $mail_config->{bcc};
    my $data_list_ref = $mail_config->{data};

    unless(defined($content_type) and ($content_type ne '')){
	$content_type = 'text/plain';
    }

    # to送信リスト作成
    my @mail_list;
    for(0 .. $#$mail_list_ref){
	my $tmp_to = '"'.$mail_list_ref->[$_]->{name}
	    .'" <'.$mail_list_ref->[$_]->{addr}.'>';
	push(@mail_list,$tmp_to);
    }

    ## 添付ファイル作成
    # boundaryランダム生成
    my @rand_chars = (0..9,'a'..'z','A'..'Z');
    my $bound = '';
    for(1 .. 32){$bound .= $rand_chars[rand($#rand_chars)]};
    # ヘッダ生成
    my $mime_header = '';
    my $attachment = '';
    if(defined($data_list_ref) and ($data_list_ref ne '')){
	$mime_header = "MIME-Version: 1.0\nContent-Type: multipart/mixed; boundary=\"$bound\"\n"
	    ."Content-Transfer-Encoding: Base64\n";
    }

    # 送信データ作成
    my $header_mailfrom = '"'.$from_addr->{name}.'" <'.$from_addr->{addr}.'>';
    my $mailto = join(',',@mail_list);

    encode('utf8',$mailto);
    encode('utf8',$header_mailfrom);
    encode('utf8',$content);
    encode('utf8',$subject);
    # メール本文ヘッダー生成
    my $header_from_addr = '';
    if(defined($from_addr) and ($from_addr ne '')){
	$header_from_addr = "From: $header_mailfrom\n";
    }
    my $header_subject = '';
    if(defined($subject) and ($subject ne '')){
	$header_subject = "Subject: $subject\n";
    }
    my $header_to_addr = '';
    if(defined($mail_list_ref) and ($mail_list_ref ne '')){
	$header_to_addr = "To: $mailto\n";
    }

    # cc送信リスト作成
    my @mail_list_cc;
    my $mailcc = '';
    if(defined($mail_list_ref_cc) and ($mail_list_ref_cc ne '')){
	for(0 .. $#$mail_list_ref_cc){
	    my $tmp_to = '"'.$mail_list_ref_cc->[$_]->{name}
		.'" <'.$mail_list_ref_cc->[$_]->{addr}.'>';
	    push(@mail_list_cc,$tmp_to);
	}
	$mailcc = "Cc: ".join(',',@mail_list_cc)."\n";
	encode('utf8',$mailcc);
    }

    # bcc送信リスト作成
    my @mail_list_bcc;
    my $mailbcc = '';
    if(defined($mail_list_ref_bcc) and ($mail_list_ref_bcc ne '')){
	for(0 .. $#$mail_list_ref_bcc){
	    my $tmp_to = '"'.$mail_list_ref_bcc->[$_]->{name}
		.'" <'.$mail_list_ref_bcc->[$_]->{addr}.'>';
	    push(@mail_list_bcc,$tmp_to);
	}
	$mailbcc = "Bcc: ".join(',',@mail_list_bcc)."\n";
	encode('utf8',$mailbcc);
    }

    my $smtp = Net::SMTP->new($mta_addr);
    my $mail_head =
	$mime_header.
	$header_from_addr.
	$header_to_addr.
	$mailcc.
	$mailbcc.
	$header_subject.
	"Mime-Version: 1.0\n";
    if(defined($data_list_ref)){
	$mail_head .= "--$bound\n";
    }
    $mail_head .= "Content-Type: $content_type; charset=\"UTF-8\"\n\n";
    $smtp->mail($from_addr->{addr});
    $smtp->to(@mail_list);
    if(defined($mail_list_ref_cc)){$smtp->cc(@mail_list_cc)}
    if(defined($mail_list_ref_bcc)){$smtp->cc(@mail_list_bcc)}
    $smtp->data();
    $smtp->datasend($mail_head);
    $smtp->datasend($content);
    if(defined($data_list_ref)){
	for(0 .. $#$data_list_ref){
	    my $attache_filename = $data_list_ref->[$_]->{filename};
	    encode('utf8',$attache_filename);
	    $smtp->datasend("\n");
	    $smtp->datasend("--$bound\n");
	    $smtp->datasend("Content-Transfer-Encoding: Base64\n");
	    $smtp->datasend("Content-Type: application/octet-stream\n");
	    $smtp->datasend("Content-Disposition: attachment; filename=\"$attache_filename\"\n\n");
	    $smtp->datasend($data_list_ref->[$_]->{data}."\n");
	}
	$smtp->datasend("--$bound--\n");
    }
    $smtp->datasend();
    $smtp->quit;
}

__PACKAGE__->meta->make_immutable();

1;


lib/YCGL/Lite/Parallel.pm
package YCGL::Lite::Parallel;
use strict;
use warnings;
use utf8;
use Mouse;
use Parallel::ForkManager;
use YCGL::Lite::HTTPClient;
use YCGL::Lite::DataConverter;
use B::Deparse;

sub do_with_result {
    my $self = shift;
    my $data = shift;
    my $code_ref = shift;
    my $max_proc = shift;
    my $result;
    my $pm = Parallel::ForkManager->new($max_proc);
    $pm->run_on_finish(
	sub {
	    my ($pid, $exit_code, $ident, $exit_signal, $core_dump, $data_structure) = @_;
	    if (defined $data_structure) {
		$result->[$data_structure->{id}] = $data_structure->{result};
	    }
	}
       );
    for(my $k=0; $k <= $#$data; $k++){
	$pm->start and next;
	my $result_chil = $code_ref->($data->[$k]);
	my $result_with_id = {id => $k, result => $result_chil};
	$pm->finish(0,$result_with_id);
    }
    $pm->wait_all_children;
    return($result);
}

sub do_without_result {
    my $self = shift;
    my $data = shift;
    my $code_ref = shift;
    my $max_proc = shift;
    my $pm = Parallel::ForkManager->new($max_proc);
    for(my $k=0; $k <= $#$data; $k++){
	$pm->start and next;
	$code_ref->($data->[$k]);
	$pm->finish;
    }
    $pm->wait_all_children;
    return;
}

sub map_reduce {
    my $self = shift;
    my $data = shift;
    my $mapper_ref = shift;
    my $reducer_ref = shift;
    my $max_proc = shift;
    my $options = shift;
    my $remote_flg = 0;
    if(defined($options) and defined($options->{remote})){
	$remote_flg = $options->{remote};
    }
    my $result;
    my $pm = Parallel::ForkManager->new($max_proc);
    $pm->run_on_finish(
	sub {
	    my ($pid, $exit_code, $ident, $exit_signal, $core_dump, $data_structure) = @_;
	    if (defined $data_structure) {
		$result->[$data_structure->{id}] = $data_structure->{result};
	    }
	}
       );
    if($remote_flg == 1){
	for(my $k=0; $k <= $#$data; $k++){
	    $pm->start and next;
	    my $stringified_code = B::Deparse->new->coderef2text($mapper_ref);
	    my $payload = YCGL::Lite::DataConverter->perl_to_json(
		{
		    data => $data->[$k]->[0],
		    code => $stringified_code
		   }
	       );
	    my $result_chil_from_remote = YCGL::Lite::HTTPClient->post_content(
		$data->[$k]->[1],
		'application/json',
		$payload
	       );
	    my $result_chil = YCGL::Lite::DataConverter->json_to_perl($result_chil_from_remote);
	    
	    my $result_with_id = {id => $k, result => $result_chil->{result}};
	    $pm->finish(0,$result_with_id);
	}
    }else{
	for(my $k=0; $k <= $#$data; $k++){
	    $pm->start and next;
	    my $result_chil = $mapper_ref->($data->[$k]);
	    my $result_with_id = {id => $k, result => $result_chil};
	    $pm->finish(0,$result_with_id);
	}
    }
    $pm->wait_all_children;
    return($reducer_ref->($result));
}


__PACKAGE__->meta->make_immutable();

1;


lib/YCGL/Lite/HTTPClient.pm
package YCGL::Lite::HTTPClient;
use strict;
use warnings;
use utf8;
use Mouse;
use WWW::Mechanize;

sub get{
    my $self = shift;
    my $url = shift;
    my $options = shift;
    my $ssl_opt = 1;
    if(defined($options) and defined($options->{ssl_verify_hostname})){
	$ssl_opt = $options->{ssl_verify_hostname};
    }
    my $mech = WWW::Mechanize->new(
	ssl_opts => {
	    verify_hostname => $ssl_opt
	   }
       );

    $mech->get($url) or die "$@\n";
    my $res = $mech->content();
    return $res;
}

sub post{
    my $self = shift;
    my $url = shift;
    my $params = shift;
    my $options = shift;
    my $ssl_opt = 1;
    if(defined($options) and defined($options->{ssl_verify_hostname})){
	$ssl_opt = $options->{ssl_verify_hostname};
    }
    my $mech = WWW::Mechanize->new(
	ssl_opts => {
	    verify_hostname => $ssl_opt
	   }
       );

    $mech->post($url,$params) or die "$@\n";
    my $res = $mech->content();
    return $res;
}

sub post_content {
    my $self = shift;
    my $url = shift;
    my $content_type = shift;
    my $data = shift;
    my $options = shift;
    my $ssl_opt = 1;
    if(defined($options) and defined($options->{ssl_verify_hostname})){
	$ssl_opt = $options->{ssl_verify_hostname};
    }
    my $mech = WWW::Mechanize->new(
	ssl_opts => {
	    verify_hostname => $ssl_opt
	   }
       );
    $mech->post($url,'Content-Type' => $content_type, Content => $data);
    my $res = $mech->content();
    return $res;
}

__PACKAGE__->meta->make_immutable();

1;


lib/YCGL/Lite/DataConverter.pm
package YCGL::Lite::DataConverter;
use strict;
use warnings;
use utf8;
use Mouse;
use JSON;
use Text::CSV;
use XML::Simple;
use Data::MessagePack;


sub perl_to_json {
    my $self = shift;
    my $data = shift;
    my $json_text = JSON::encode_json $data;
    return($json_text);
}

sub json_to_perl {
    my $self = shift;
    my $json = shift;
    my $perl_data = JSON::decode_json $json;
    return($perl_data);
}

sub perl_to_csv {
    my $self = shift;
    my $data = shift;
    my $options = shift;
    my $EOL = "\n";
    if(defined($options) and defined($options->{eol})){
	$EOL = $options->{eol};
    }
    my $csv = Text::CSV->new({binary => 1});
    $csv->eol($EOL);
    my $serialized = '';
    open my $fh, '>', \$serialized;
    for(my $k=0; $k <= $#$data; $k++){
	$csv->print($fh,$data->[$k]);
    }
    close($fh);
    return($serialized);
}

sub csv_to_perl {
    my $self = shift;
    my $text = shift;
    my $csv = Text::CSV->new({binary => 1});
    open my $fh, '<', \$text;
    my $rows = $csv->getline_all($fh);
    close($fh);
    return($rows);
}

sub csvfile_to_perl {
    my $self = shift;
    my $file = shift;
    my $csv = Text::CSV->new({binary => 1});
    open my $fh, '<', $file;
    my $rows = $csv->getline_all($fh);
    close($fh);
    return($rows);
}

sub perl_to_csvfile {
    my $self = shift;
    my $data = shift;
    my $filename = shift;
    my $options = shift;
    my $EOL = "\n";
    if(defined($options) and defined($options->{eol})){
	$EOL = $options->{eol};
    }
    my $csv = Text::CSV->new({binary => 1});
    $csv->eol($EOL);
    open my $fh, '>', $filename;
    for(my $k=0; $k <= $#$data; $k++){
	$csv->print($fh,$data->[$k]);
    }
    close($fh);
}

sub perl_to_xml {
    my $self = shift;
    my $records = shift;
    my $options = shift;
    my $xml = XML::Simple->new(
	ForceArray => $options->{ForceArray},
	RootName => $options->{RootName},
	XMLDecl => $options->{XMLDecl}
       );
    my $xml_out = $xml->XMLout($records);
    return($xml_out);
}

sub xml_to_perl {
    my $self = shift;
    my $xml_text = shift;
    my $options = shift;
    my $xml = XML::Simple->new(
	ForceArray => $options->{ForceArray},
	RootName => $options->{RootName},
	XMLDecl => $options->{XMLDecl}
       );
    my $perl_from_xml = $xml->XMLin($xml_text);
    return($perl_from_xml);
}

sub perl_to_msgpack {
    my $self = shift;
    my $data = shift;
    my $msgpack = Data::MessagePack->new();
    my $packed = $msgpack->pack($data);
    return($packed);
}

sub msgpack_to_perl {
    my $self = shift;
    my $msg_text = shift;
    my $msgpack = Data::MessagePack->new();
    my $unpacked = $msgpack->unpack($msg_text);
    return($unpacked);
}

__PACKAGE__->meta->make_immutable();

1;


lib/YCGL/Lite.pm
package YCGL::Lite;
use 5.008001;
use strict;
use warnings;
use YCGL::Lite::Times;
use YCGL::Lite::Mailer;
use YCGL::Lite::HTTPClient;
use YCGL::Lite::DataConverter;
use YCGL::Lite::Parallel;
use YCGL::Lite::Plack;
use Mouse;

our $VERSION = "0.01";

has 'times' => (is => 'ro', default => sub { YCGL::Lite::Times->new });
has 'mail' => (is => 'ro', default => sub { YCGL::Lite::Mailer->new });
has 'http_client' => (is => 'ro', default => sub { YCGL::Lite::HTTPClient->new });
has 'data_conv' => (is => 'ro', default => sub { YCGL::Lite::DataConverter->new });
has 'parallel' => (is => 'ro', default => sub { YCGL::Lite::Parallel->new });
has 'plack' => (is => 'ro', default => sub { YCGL::Lite::Plack->new });


__PACKAGE__->meta->make_immutable();

1;
__END__

=encoding utf-8

=head1 NAME

YCGL::Lite - Yokoda Common General Library lite.

=head1 SYNOPSIS

    use strict;
    use warnings;
    use feature qw(say);
    use YCGL::Lite;

    my $ycgl = YCGL::Lite->new();

    ## Current timestamp
    my $current_timestamp = $ycgl->times->current_timestamp();
    say $current_timestamp; # 2016-01-01 00:00:00
    my $today = $ycgl->times->today();
    say $today; # 2016-01-01

    my $date_one_month_ago = $ycgl->times->date_add($today,0,-1,0);
    say $date_one_month_ago; # 2015-12-01
    my $date_diff = $ycgl->times->date_diff($today,'2016-01-05');
    say $date_diff; # 4

    ## Mail
    my $mail_data = {
       to => [{name => 'to1',addr =>'to1@example.com'}],
       cc => [{name => 'cc1', addr => 'cc1@example.com'}],
       bcc => [{name => 'bcc', addr => 'bcc1@example.com'}],
       from => {name => 'myname', addr => 'myname@example.com'},
       mta_addr => '127.0.0.1',
       subject => 'mail example.',
       content => 'This is example.',
       content_type => 'text/plain'
    };

    $ycgl->mail->send_mail($mail_data);

    ## HTTP, HTTPS client
    my $result_http = $ycgl->http_client->get('http://example.com');
    my $result_https = $ycgl->http_client->post('https://example.com',{username => 'myname', password => 'password'});

    ## Data interchange
    # JSON
    my $json_text = $ycgl->data_conv->perl_to_json({name => 'myname', age => 20});
    my $perl_from_json = $ycgl->data_conv->json_to_perl($json_text);

    # CSV
    my $csv_text = $ycgl->data_conv->perl_to_csv([['name','age'],['myname',20],['yourname',21]]);
    my $perl_from_csv = $ycgl->data_conv->csv_to_perl($csv_text);
    my $perl_from_csvfile = $ycgl->data_conv->csvfile_to_perl('./mycsvfile.csv');
    $ycgl->data_conv->perl_to_csvfile($perl_from_csvfile,'./mycsvoutput.csv');

    # XML
    my $xml_records = {
      meta_data => [{source_name => ['MYSOURCE']}],
      records =>
      [
        {
          data_seq => 1,
          pname => 'myname',
          age => 20
        },
        {
          data_seq => 2,
          pname => 'yourname',
          age => 21
        }
      ]
    };
    my $xml_text = $ycgl->data_conv->perl_to_xml(
      $xml_records,
      {ForceArray => 1,RootName => 'ROOTNAME', XMLDecl => 1} # option
    );
    my $perl_from_xml = $ycgl->data_conv->xml_to_perl($xml_text);

    # MessagePack
    my $msgpack_text = $ycgl->data_conv->perl_to_msgpack({name => 'myname', age => 20});
    my $perl_from_msgpack = $ycgl->data_conv->msgpack_to_perl($msgpack_text);

    ## Parallel Processing
    my $data_parallel_1 = [{val_1 => 1, val_2 => 2},{val_1 => 10, val_2 => 20}];
    my $my_sub_1 = sub {my $inputs = shift; return($inputs{val_1} + $inputs{val_2});};
    my @results = $ycgl->parallel->do_with_result($data_parallel_1, $my_sub_1, 10);

    my $data_parallel_2 = ['http://example.com/1','http://example.com/2'];
    my $my_sub_2 = sub {my $url = shift; $ycgl->http_client->get($url);};
    $ycgl->parallel->do_without_result($data_parallel_2, $my_sub_2, 20);

    ## MapReduce
    # generating data
    my $data_map_reduce;
    for(0 .. 20){
        my $tmp_data;
        for(0 .. 10000){
            push(@$tmp_data,rand(10000));
        }
        push(@$data_map_reduce,$tmp_data);
    }

    # mapper code
    my $mapper = sub {
        my $input = shift;
        my $sum = 0;
        my $num = $#$input + 1;
        for(0 .. $#$input){
            $sum += $input->[$_];
        }
        my $avg = $sum / $num;
        return({avg => $avg, sum => $sum});
    };

    # reducer code
    my $reducer = sub {
        my $input = shift;
        my $sum = 0;
        my $avg = 0;
        my $num = $#$input + 1;
        for(0 .. $#$input){
            $sum += $input->[$_]->{sum};
            $avg += $input->[$_]->{avg};
        }
        $avg = $avg / $num;
        return({avg => $avg, sum => $sum});
    };

    # do MapReduce
    my $result = $ycgl->parallel->map_reduce(
        $data,
        $mapper,
        $reducer,
        10
       );
    print "SUM: $result->{sum}\nAVG: $result->{avg}\n";

    ## Plack Server
    # static file server
    my $public_dir = './public/';
    $ycgl->plack->plackup_static($public_dir);

    # eval server (dangerous)
    my $eval_path = '/mysecret_eval_path';
    $ycgl->plack->plackup_eval($eval_path);

    my $data_map_reduce_real;
    for(0 .. 2){
        my $tmp_data;
        for(0 .. 10000){
            push(@$tmp_data,rand(10000));
        }
        push(@$data_map_reduce_real,[$tmp_data,'http://localhost:5000/mysecret_eval_path']);
    }

    # MapReduce by using remote eval servers
    my $real_mapr_result = $ycgl->parallel->map_reduce(
        $data_map_reduce_real,
        $mapper,
        $reducer,
        3,
        {remote => 1}
        );


=head1 DESCRIPTION

YCGL::Lite has commonly and generally used functions.


=head1 LICENSE

Copyright (C) Toshiaki Yokoda.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Toshiaki Yokoda E<lt>adokoy001@gmail.comE<gt>

=cut

